File: old/core/aggregators/reddit/content.py
"""Reddit content building utilities."""

import html
import logging
from typing import List

from ..exceptions import ArticleSkipError
from .comments import fetch_post_comments, format_comment_html
from .markdown import convert_reddit_markdown, safe_img_html, safe_link_html
from .types import RedditPostData
from .urls import decode_html_entities_in_url, fix_reddit_media_url

logger = logging.getLogger(__name__)


def build_post_content(
    post: RedditPostData,
    comment_limit: int,
    subreddit: str,
    user_id: int,
    is_cross_post: bool = False,
) -> str:
    """
    Build post content with comments.

    Args:
        post: RedditPostData instance
        comment_limit: Number of comments to fetch
        subreddit: Subreddit name
        user_id: User ID for authentication
        is_cross_post: Whether this is a cross-post

    Returns:
        HTML content string
    """
    content_parts: List[str] = []

    # Add selftext part
    if post.selftext:
        selftext_html = convert_reddit_markdown(post.selftext)
        content_parts.append(f"<div>{selftext_html}</div>")

    # Add gallery media
    _add_gallery_media(post, content_parts)

    # Add link media
    _add_link_media(post, content_parts, is_cross_post)

    # Add comments section
    _add_comments_section(post, comment_limit, subreddit, user_id, content_parts)

    return "".join(content_parts)


def _add_selftext_part(post: RedditPostData, content_parts: List[str]) -> None:
    """Add selftext part to content."""
    if post.selftext:
        selftext_html = convert_reddit_markdown(post.selftext)
        content_parts.append(f"<div>{selftext_html}</div>")


def _process_gallery_item(item: dict, post: RedditPostData) -> str | None:
    """Process a single gallery item."""
    media_id = item.get("media_id")
    if not media_id:
        return None

    media_info = post.media_metadata.get(media_id) if post.media_metadata else None
    if not media_info:
        return None

    is_animated = media_info.get("e") == "AnimatedImage"
    media_url = None
    if is_animated:
        media_url = media_info.get("s", {}).get("gif") or media_info.get("s", {}).get("mp4")
    elif media_info.get("e") == "Image":
        media_url = media_info.get("s", {}).get("u")

    if not media_url:
        return None

    fixed_url = fix_reddit_media_url(decode_html_entities_in_url(media_url))
    caption = item.get("caption", "")
    alt = "Gallery image"
    if caption:
        alt = caption
    elif is_animated:
        alt = "Animated GIF"

    # fixed_url and alt/caption are attacker-reachable (gallery data comes
    # from the Reddit API for a post an arbitrary user submitted), so the
    # image is built through safe_img_html rather than interpolated raw: an
    # unsafe scheme (javascript:/data:) skips the image entirely instead of
    # rendering it, and a quote in either value can't break out of its
    # attribute.
    img_html = safe_img_html(fixed_url, alt)
    if not img_html:
        return None

    if caption:
        return f"<figure>{img_html}<figcaption>{html.escape(alt, quote=True)}</figcaption></figure>"
    return f"<p>{img_html}</p>"


def _add_gallery_media(post: RedditPostData, content_parts: List[str]) -> None:
    """Add gallery media to content."""
    if not post.is_gallery or not post.media_metadata or not post.gallery_data:
        return

    items = post.gallery_data.get("items", [])
    for item in items:
        html = _process_gallery_item(item, post)
        if html:
            content_parts.append(html)


def _add_link_media(post: RedditPostData, content_parts: List[str], is_cross_post: bool) -> None:
    """Add link media to content."""
    if not post.url or post.is_gallery:
        return

    url = decode_html_entities_in_url(post.url)

    # Try media handlers in order
    if _process_link_media(post, url, content_parts):
        return

    # Fallback link. url is the post's own submitted link -- attacker-
    # reachable -- so it goes through safe_link_html rather than being
    # interpolated raw: an unsafe scheme renders as bare text instead of a
    # live href, and a quote in the URL can't break out of the attribute.
    if not is_cross_post and not post.is_self:
        content_parts.append(f"<p>{safe_link_html(url, url)}</p>")


def _process_link_media(post: RedditPostData, url: str, content_parts: List[str]) -> bool:
    """Process link media by delegating to appropriate handler."""
    url_lower = url.lower()

    # Handle GIF media
    if url_lower.endswith((".gif", ".gifv")):
        from .images import extract_animated_gif_url

        gif_url = extract_animated_gif_url(post) or (
            url[:-1] if url_lower.endswith(".gifv") else url
        )
        fixed_url = fix_reddit_media_url(gif_url)
        img_html = safe_img_html(fixed_url, "Animated GIF")
        if img_html:
            content_parts.append(f"<p>{img_html}</p>")
        return True

    # Handle direct image media
    is_image = (
        any(ext in url_lower for ext in [".jpg", ".jpeg", ".png", ".webp"])
        or "i.redd.it" in url_lower
    )
    if is_image:
        fixed_url = fix_reddit_media_url(url)
        if fixed_url:
            content_parts.append(f"<p>{safe_link_html(fixed_url, fixed_url)}</p>")
        return True

    # Handle video media (Reddit videos and YouTube)
    if "v.redd.it" in url_lower:
        # Note: v.redd.it video links and thumbnails are now handled in the header
        # via aggregator.py process_content, so we don't add them to the body here.
        return True

    if "youtube.com" in url_lower or "youtu.be" in url_lower:
        content_parts.append(f"<p>{safe_link_html(url, '▶ View Video on YouTube')}</p>")
        return True

    # Handle Twitter/X links (embed is rendered in the header, not in body)
    return "twitter.com" in url_lower or "x.com" in url_lower


def _add_comments_section(
    post: RedditPostData,
    comment_limit: int,
    subreddit: str,
    user_id: int,
    content_parts: List[str],
) -> None:
    """Add comments section to content."""
    decoded_permalink = decode_html_entities_in_url(post.permalink)
    permalink = f"https://reddit.com{decoded_permalink}"
    comment_section_parts = [f"<h3>{safe_link_html(permalink, 'Comments')}</h3>"]

    if comment_limit > 0:
        try:
            comments = fetch_post_comments(subreddit, post.id, comment_limit, user_id)
            if comments:
                comment_htmls = [format_comment_html(comment) for comment in comments]
                comment_section_parts.append("".join(comment_htmls))
            else:
                comment_section_parts.append("<p><em>No comments yet.</em></p>")
        except ArticleSkipError:
            raise
        except Exception as e:
            logger.warning(f"Failed to fetch comments: {e}")
            comment_section_parts.append("<p><em>Comments unavailable.</em></p>")
    else:
        comment_section_parts.append("<p><em>Comments disabled.</em></p>")

    content_parts.append(f"<section>{''.join(comment_section_parts)}</section>")

File: src/lib/aggregators/sites/reddit/content.ts
/**
 * Reddit content building utilities.
 *
 * Ported from old/core/aggregators/reddit/content.py.
 */

import { ArticleSkipError } from "../../errors";
import type { ChromeLabels } from "../../chrome-labels";
import { fetchPostComments, formatCommentHtml } from "./comments";
import { extractAnimatedGifUrl, extractGiphyGifUrl } from "./images";
import { convertRedditMarkdown, escapeHtml, safeImgHtml, safeLinkHtml } from "./markdown";
import { RedditComment, RedditGalleryItem, RedditPostData } from "./types";
import { decodeHtmlEntitiesInUrl, fixRedditMediaUrl } from "./urls";

/**
 * What a crosspost is, spelled out for the reader. A crosspost's article body
 * is built from the *original* post -- its title, selftext, media, comments
 * and permalink all belong to the subreddit it was first submitted to -- so
 * without this nothing in the finished article says the post reached the feed
 * by way of a crosspost, nor which subreddit it actually came from. That was
 * true of the retired Django implementation too: `is_cross_post` only ever
 * suppressed the bare link `addLinkMedia()` would otherwise append (the
 * crosspost's `url` is the original post, so the link would point the reader
 * back at the article they are already reading).
 *
 * The subreddit the crosspost itself appeared in is deliberately *not* part of
 * this: it is the feed's own subreddit, which the reader already knows -- the
 * one thing the article does not carry is where the post came *from*.
 * `originalSubreddit` is `""` when Reddit's `crosspost_parent_list` entry
 * carries no `subreddit` (rare), which is distinct from a null attribution
 * meaning "not a crosspost at all".
 */
export interface CrosspostAttribution {
  originalSubreddit: string;
}

/**
 * The one-line "Crosspost: r/from" notice that opens a crosspost's body, the
 * subreddit name linking to that subreddit. Exported because
 * `extractContent()`'s JSON branch in `./aggregator.ts` builds its content
 * without going through `buildPostContent()` and has to emit the same notice
 * rather than a second version of it.
 *
 * With no origin known the notice degrades to the bare word, which still
 * answers the question the whole notice exists for -- is this a crosspost.
 */
export function buildCrosspostNoticeHtml(
  crosspost: CrosspostAttribution,
  labels: ChromeLabels,
): string {
  const origin = crosspost.originalSubreddit
    ? `: ${safeLinkHtml(
        `https://reddit.com/r/${crosspost.originalSubreddit}`,
        `r/${crosspost.originalSubreddit}`,
      )}`
    : "";

  return `<p><em>${escapeHtml(labels.crosspost)}${origin}</em></p>`;
}

/**
 * `buildPostContent()`'s return: the post's own body, and its comment section
 * kept separate rather than concatenated in. Callers that fingerprint or
 * render this content must pass `comments` through `formatArticleContent()`'s
 * `commentsContent` parameter -- never splice it into `body` themselves -- or
 * `articleContentHash()`'s comment exclusion
 * (`src/lib/aggregators/content-hash.ts`) cannot find it and a busy thread
 * rewrites the row on every aggregation cycle. See `ARTICLE_COMMENTS_CLASS` in
 * `../../extract/format` for the one place that wrapper is written.
 */
export interface RedditPostContent {
  body: string;
  comments: string | null;
}

export async function buildPostContent(
  post: RedditPostData,
  commentLimit: number,
  subreddit: string,
  labels: ChromeLabels,
  userId?: number | string | null,
  crosspost: CrosspostAttribution | null = null,
  commentsList?: RedditComment[],
): Promise<RedditPostContent> {
  const contentParts: string[] = [];

  // 1. Crosspost notice -- first, so the reader knows what they are looking at
  // before the original post's own body starts.
  if (crosspost) {
    contentParts.push(buildCrosspostNoticeHtml(crosspost, labels));
  }

  // 2. Selftext
  if (post.selftext) {
    const selftextHtml = convertRedditMarkdown(post.selftext);
    contentParts.push(`<div>${selftextHtml}</div>`);
  }

  // 3. Gallery media
  addGalleryMedia(post, contentParts);

  // 4. Link media
  addLinkMedia(post, contentParts, Boolean(crosspost), labels);

  // 5. Comments section -- built separately, not pushed into contentParts. See
  // `RedditPostContent` above for why.
  const comments = await buildCommentsSection(
    post,
    commentLimit,
    subreddit,
    userId,
    labels,
    commentsList,
  );

  return { body: contentParts.join(""), comments };
}

function processGalleryItem(item: RedditGalleryItem, post: RedditPostData): string | null {
  const mediaId = item.media_id;
  if (!mediaId || !post.media_metadata) return null;

  const mediaInfo = post.media_metadata[mediaId];
  if (!mediaInfo) return null;

  const isAnimated = mediaInfo.e === "AnimatedImage";
  let mediaUrl: string | null = null;
  if (isAnimated) {
    mediaUrl = mediaInfo.s?.gif || mediaInfo.s?.mp4 || null;
  } else if (mediaInfo.e === "Image") {
    mediaUrl = mediaInfo.s?.u || null;
  }

  if (!mediaUrl) return null;

  const fixedUrl = fixRedditMediaUrl(decodeHtmlEntitiesInUrl(mediaUrl));
  const caption = item.caption || "";
  const alt = caption || (isAnimated ? "Animated GIF" : "Gallery image");

  const imgHtml = safeImgHtml(fixedUrl, alt);
  if (!imgHtml) return null;

  if (caption) {
    return `<figure>${imgHtml}<figcaption>${escapeHtml(alt)}</figcaption></figure>`;
  }
  return `<p>${imgHtml}</p>`;
}

function addGalleryMedia(post: RedditPostData, contentParts: string[]): void {
  if (!post.is_gallery || !post.media_metadata || !post.gallery_data) {
    return;
  }

  const items = post.gallery_data.items || [];
  for (const item of items) {
    const html = processGalleryItem(item, post);
    if (html) {
      contentParts.push(html);
    }
  }
}

function addLinkMedia(
  post: RedditPostData,
  contentParts: string[],
  isCrossPost: boolean,
  labels: ChromeLabels,
): void {
  if (!post.url || post.is_gallery) return;

  const url = decodeHtmlEntitiesInUrl(post.url);

  if (processLinkMedia(post, url, contentParts, labels)) {
    return;
  }

  if (!isCrossPost && !post.is_self) {
    contentParts.push(`<p>${safeLinkHtml(url, url)}</p>`);
  }
}

function processLinkMedia(
  post: RedditPostData,
  url: string,
  contentParts: string[],
  labels: ChromeLabels,
): boolean {
  const urlLower = url.toLowerCase();

  const giphyUrl = extractGiphyGifUrl(url);
  if (giphyUrl) {
    const imgHtml = safeImgHtml(giphyUrl, "Giphy");
    if (imgHtml) contentParts.push(`<p>${imgHtml}</p>`);
    return true;
  }

  // GIF media
  if (urlLower.endsWith(".gif") || urlLower.endsWith(".gifv")) {
    const gifUrl =
      extractAnimatedGifUrl(post) || (urlLower.endsWith(".gifv") ? url.slice(0, -1) : url);
    const fixedUrl = fixRedditMediaUrl(gifUrl);
    const imgHtml = safeImgHtml(fixedUrl, "Animated GIF");
    if (imgHtml) {
      contentParts.push(`<p>${imgHtml}</p>`);
    }
    return true;
  }

  // Direct image media
  const isImage =
    [".jpg", ".jpeg", ".png", ".webp"].some((ext) => urlLower.includes(ext)) ||
    urlLower.includes("i.redd.it");
  if (isImage) {
    const fixedUrl = fixRedditMediaUrl(url);
    if (fixedUrl) {
      contentParts.push(`<p>${safeLinkHtml(fixedUrl, fixedUrl)}</p>`);
    }
    return true;
  }

  // Video media
  if (urlLower.includes("v.redd.it")) {
    return true;
  }

  if (urlLower.includes("youtube.com") || urlLower.includes("youtu.be")) {
    contentParts.push(`<p>${safeLinkHtml(url, labels.viewVideoOnYoutube)}</p>`);
    return true;
  }

  // Twitter/X links
  return urlLower.includes("twitter.com") || urlLower.includes("x.com");
}

/**
 * Builds the comment section's own inner markup -- the caller wraps it (or
 * doesn't); this function never wraps itself in `<section>`, so the only
 * `ARTICLE_COMMENTS_CLASS` wrapper in the whole pipeline stays the one
 * `formatArticleContent()` writes.
 */
async function buildCommentsSection(
  post: RedditPostData,
  commentLimit: number,
  subreddit: string,
  userId: number | string | null | undefined,
  labels: ChromeLabels,
  providedComments?: RedditComment[],
): Promise<string> {
  const decodedPermalink = decodeHtmlEntitiesInUrl(post.permalink);
  const permalink = `https://reddit.com${decodedPermalink}`;
  const commentSectionParts: string[] = [`<h3>${safeLinkHtml(permalink, labels.comments)}</h3>`];

  if (commentLimit > 0) {
    try {
      const comments =
        providedComments !== undefined
          ? providedComments
          : await fetchPostComments(subreddit, post.id, commentLimit, userId);

      if (comments && comments.length > 0) {
        const commentHtmls = comments.map((comment) => formatCommentHtml(comment, labels));
        commentSectionParts.push(commentHtmls.join(""));
      } else {
        commentSectionParts.push(`<p><em>${labels.noCommentsYet}</em></p>`);
      }
    } catch (err) {
      // A 403/404 from the comments endpoint means the post itself is private,
      // removed or gone -- `fetchPostComments()` reports that as an
      // `ArticleSkipError` and the caller drops the article. Swallowing it here
      // would silently reinstate the bug that fix by degrading a skipped post
      // into one whose body says "Comments unavailable." Production always
      // pre-fetches (`aggregator.ts` passes `commentsList`), so this path is
      // reachable only from a future caller that does not -- which is exactly
      // when the guard has to already be here.
      if (err instanceof ArticleSkipError) throw err;
      commentSectionParts.push(`<p><em>${labels.commentsUnavailable}</em></p>`);
    }
  } else {
    commentSectionParts.push(`<p><em>${labels.commentsDisabled}</em></p>`);
  }

  return commentSectionParts.join("");
}
